### ISSUE TITLE
Recipe training phase exceeds 60-second threshold

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#835]
Phase: training

Diagnostic: elapsed_seconds:63 exceeded threshold_seconds:60 during the public Recipe run.

Public entrypoint: recipe (binary) invoked as `cargo run --bin recipe -- <reproduction.rs>` with the `recipe.train()` API path on backend `nv4`, using data `data/numeric/sample_subfolders`.

Evidence: process snapshot captured after 63 seconds shows the parent `recipe` process in `do_wait` (60S), its `recipe-script-2` child in `poll_schedule_timeout.constprop.` (60S), and a defunct `recipe-script-2` process (19Zl), while the reproduction's `phase("training")` call had not returned. The failing phase was `training` with `epochs(1)`, `optimizer(adamw)`, `lr(0.001)`, `seed(17)`, `log(all)`, and `fp(32)` on a model composed of `layer(16)`, `moe(1, [layer(8), layer(24), layer(8), layer(24), layer(8), layer(24), layer(8), layer(24)])`, and `layer(1)` with `loss(mse)`.

Rationale: hypothesis — the `recipe.train()` call on the MoE model is spending its runtime in the training loop or its scheduling/wait path on the `nv4` backend; the reproduction is not performed and stability is `false`, so no fix is proposed.

## Machine provenance

- Discoverer: Recipe composition traversal
- Failure fingerprint: `36ae603eb764ad60`
- Recipe source commit: `e374b525162643338ab16d237e6f1257240a5670`
- Embedded Recipe `recipe.rs` blob: `7647f63c4134b7807252948983c37c9cad78193d`
- Source identity: The packet captures the exact working-tree Recipe source independently of the Recipe Machine commit.
- Classifier provider: `openrouter`
- Classifier model: `openrouter/free`
- Classifier effort: `default`

### Classifier decision

```json
{"verdict":"new","title":"Recipe training phase exceeds 60-second threshold","body":"Phase: training\n\nDiagnostic: elapsed_seconds:63 exceeded threshold_seconds:60 during the public Recipe run.\n\nPublic entrypoint: recipe (binary) invoked as `cargo run --bin recipe -- <reproduction.rs>` with the `recipe.train()` API path on backend `nv4`, using data `data/numeric/sample_subfolders`.\n\nEvidence: process snapshot captured after 63 seconds shows the parent `recipe` process in `do_wait` (60S), its `recipe-script-2` child in `poll_schedule_timeout.constprop.` (60S), and a defunct `recipe-script-2` process (19Zl), while the reproduction's `phase(\"training\")` call had not returned. The failing phase was `training` with `epochs(1)`, `optimizer(adamw)`, `lr(0.001)`, `seed(17)`, `log(all)`, and `fp(32)` on a model composed of `layer(16)`, `moe(1, [layer(8), layer(24), layer(8), layer(24), layer(8), layer(24), layer(8), layer(24)])`, and `layer(1)` with `loss(mse)`.\n\nRationale: hypothesis — the `recipe.train()` call on the MoE model is spending its runtime in the training loop or its scheduling/wait path on the `nv4` backend; the reproduction is not performed and stability is `false`, so no fix is proposed.","rationale":"The earliest observed operation that failed was the public `recipe.train()` entrypoint during the `training` phase; the evidence shows the process remained in `do_wait`/`poll_schedule_timeout` for 63 seconds past the 60-second threshold. The hypothesis is that the MoE training loop or its scheduling path on `nv4` is consuming excessive time, but this is not proven since the reproduction was not rerun and stability is `false`."}
```

## Deterministic failure packet

kind=performance
id=36ae603eb764ad60
base=commit=e374b525162643338ab16d237e6f1257240a5670 tracked_tree=clean recipe_source=7647f63c4134b7807252948983c37c9cad78193d
data=path:"data/numeric/sample_subfolders" hash:5d3526a2d71caf6d
cursor=cursor:106199 next:106200 composition:106199
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 63 seconds
measurement=elapsed_seconds:63 threshold_seconds:60 backend:nv4 phase:training process_snapshot:captured
process_evidence:
```text
2523565 1928693 2523565      60 S    do_wait                          recipe          target/debug/recipe test.rs
2523692 2523565 2523565      60 S    poll_schedule_timeout.constprop. recipe-script-2 /home/nate/.local/share/recipe-machine/worktrees/e374b525162643338ab16d237e6f1257240a5670/target/debug/recipe-trials/recipe-trial-3711-106199/recipe-script-2523565
2529964 2523692 2523565      20 S    do_wait                          recipe          /home/nate/.local/share/recipe-machine/worktrees/e374b525162643338ab16d237e6f1257240a5670/target/debug/recipe /home/nate/.local/share/recipe-machine/worktrees/e374b525162643338ab16d237e6f1257240a5670/target/debug/recipe-trials/recipe-trial-3711-106199/reproduction.rs
2530126 2529964 2523565      19 Zl   -                                recipe-script-2 [recipe-script-2] <defunct>
```
replay=not_performed stable:false
command=cargo run --bin recipe -- /home/nate/.local/share/recipe-machine/worktrees/e374b525162643338ab16d237e6f1257240a5670/target/debug/recipe-trials/recipe-trial-3711-106199/reproduction.rs
reproduction:
```rust
use recipe::*;
use std::path::{Path, PathBuf};

fn phase(value: &str) {
	if let Some(path) = std::env::var_os("RECIPE_COMPOSITION_PHASE_PATH") {
		std::fs::write(path, value).expect("cannot write composition phase");
	}
}

fn input_width(path: &Path) -> usize {
	let text = std::fs::read_to_string(path).unwrap_or_else(|error| panic!("cannot read {}: {error}", path.display()));
	let mut shape = text.lines().find_map(|line| line.trim().strip_prefix("shape ")).expect("saved model has no input shape").split_whitespace();
	let channels = shape.next().expect("saved model has no input channels").parse::<usize>().expect("saved input channels are invalid");
	let length = shape.next().expect("saved model has no input length").parse::<usize>().expect("saved input length is invalid");
	channels.checked_mul(length).expect("saved input width overflows")
}

fn main() {
	let root = std::env::var_os("RECIPE_TRIAL_DIRECTORY").map(PathBuf::from).unwrap_or_else(std::env::temp_dir);
	let bundle = root.join(format!("composition-candidate-{}.ogdl", std::process::id()));
	phase("setup");
	let data = recipe.data("data/numeric/sample_subfolders").target("target");
	let model = recipe.model()
		.layer(16)
		.moe(1, [
			layer(8),
			layer(24),
			layer(8),
			layer(24),
			layer(8),
			layer(24),
			layer(8),
			layer(24),
		])
		.layer(1)
		.loss(mse);
	phase("training");
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(17)
		.epochs(1)
		.log(all)
		.fp(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
	phase("inference");
	let input = vec![0.0; input_width(&bundle)];
	let output = recipe.infer(&bundle, &input);
	assert!(!output.is_empty());
	assert!(output.iter().all(|value| value.is_finite()));
	std::fs::remove_file(&bundle).unwrap_or_else(|error| panic!("cannot remove {}: {error}", bundle.display()));
}
```
backend=nv4

**Comment by nm-z** [comment]
Observed evidence: public operation recipe.train().run(&model, &data) exceeded the 60-second threshold during the training phase (elapsed_seconds:60, threshold_seconds:60) on backend nv4. The model composition uses layer(16).moe(1, [layer(8), layer(24), layer(8), layer(24), layer(8), layer(24)]).layer(1).loss(mse) with data data/numeric/sample_subfolders, seed 17, and optimizer adamw. Process evidence captures the parent recipe process in do_wait and recipe-script-2 in poll_schedule_timeout.constprop, while the child reproduction executes in training.

This matches existing issue #835, which represents the identical public operation (recipe.train().run), training phase timeout against the 60-second threshold on backend nv4 with the same data, seed, recipe source commit, and MoE architecture. Differences in exact elapsed seconds or expert count are configuration details reached after the failure and do not represent a materially different earliest cause.

## Machine provenance

- Discoverer: Recipe composition traversal
- Failure fingerprint: `02fd9fc40ce5be8b`
- Recipe source commit: `e374b525162643338ab16d237e6f1257240a5670`
- Embedded Recipe `recipe.rs` blob: `7647f63c4134b7807252948983c37c9cad78193d`
- Source identity: The packet captures the exact working-tree Recipe source independently of the Recipe Machine commit.
- Classifier provider: `agy`
- Classifier model: `gemini-3.8-flash-low`
- Classifier effort: `default`

### Classifier decision

```json
{"body":"Observed evidence: public operation recipe.train().run(&model, &data) exceeded the 60-second threshold during the training phase (elapsed_seconds:60, threshold_seconds:60) on backend nv4. The model composition uses layer(16).moe(1, [layer(8), layer(24), layer(8), layer(24), layer(8), layer(24)]).layer(1).loss(mse) with data data/numeric/sample_subfolders, seed 17, and optimizer adamw. Process evidence captures the parent recipe process in do_wait and recipe-script-2 in poll_schedule_timeout.constprop, while the child reproduction executes in training.\n\nThis matches existing issue #835, which represents the identical public operation (recipe.train().run), training phase timeout against the 60-second threshold on backend nv4 with the same data, seed, recipe source commit, and MoE architecture. Differences in exact elapsed seconds or expert count are configuration details reached after the failure and do not represent a materially different earliest cause.","issue":835,"rationale":"The packet describes a deterministic performance timeout during the training phase using public operation recipe.train().run(&model, &data) on backend nv4 with data data/numeric/sample_subfolders. The earliest observable failure and implementation path match issue #835 (and related nv4 MoE training timeout issue #836). Per classification rules, 60-second training timeouts that share the earliest failing public operation, phase, dataset, and backend are duplicates; variations in cursor, elapsed time, process snapshots, or MoE expert count do not constitute a materially different cause. Therefore, this failure packet is classified as a comment on existing issue #835.","title":"Recipe training phase exceeds 60-second threshold","verdict":"comment"}
```

## Deterministic failure packet

kind=performance
id=02fd9fc40ce5be8b
base=commit=e374b525162643338ab16d237e6f1257240a5670 tracked_tree=clean recipe_source=7647f63c4134b7807252948983c37c9cad78193d
data=path:"data/numeric/sample_subfolders" hash:5d3526a2d71caf6d
cursor=cursor:105007 next:105008 composition:105007
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 threshold_seconds:60 backend:nv4 phase:training process_snapshot:captured
process_evidence:
```text
2318231 1928693 2318231      60 S    do_wait                          recipe          target/debug/recipe test.rs
2318295 2318231 2318231      60 S    poll_schedule_timeout.constprop. recipe-script-2 /home/nate/.local/share/recipe-machine/worktrees/e374b525162643338ab16d237e6f1257240a5670/target/debug/recipe-trials/recipe-trial-2519-105007/recipe-script-2318231
2324018 2318295 2318231      12 S    do_wait                          recipe          /home/nate/.local/share/recipe-machine/worktrees/e374b525162643338ab16d237e6f1257240a5670/target/debug/recipe /home/nate/.local/share/recipe-machine/worktrees/e374b525162643338ab16d237e6f1257240a5670/target/debug/recipe-trials/recipe-trial-2519-105007/reproduction.rs
2324343 2324018 2318231      11 R    -                                recipe-script-2 /home/nate/.local/share/recipe-machine/worktrees/e374b525162643338ab16d237e6f1257240a5670/target/debug/recipe-trials/recipe-trial-2519-105007/recipe-script-2324018
```
replay=not_performed stable:false
command=cargo run --bin recipe -- /home/nate/.local/share/recipe-machine/worktrees/e374b525162643338ab16d237e6f1257240a5670/target/debug/recipe-trials/recipe-trial-2519-105007/reproduction.rs
reproduction:
```rust
use recipe::*;
use std::path::{Path, PathBuf};

fn phase(value: &str) {
	if let Some(path) = std::env::var_os("RECIPE_COMPOSITION_PHASE_PATH") {
		std::fs::write(path, value).expect("cannot write composition phase");
	}
}

fn input_width(path: &Path) -> usize {
	let text = std::fs::read_to_string(path).unwrap_or_else(|error| panic!("cannot read {}: {error}", path.display()));
	let mut shape = text.lines().find_map(|line| line.trim().strip_prefix("shape ")).expect("saved model has no input shape").split_whitespace();
	let channels = shape.next().expect("saved model has no input channels").parse::<usize>().expect("saved input channels are invalid");
	let length = shape.next().expect("saved model has no input length").parse::<usize>().expect("saved input length is invalid");
	channels.checked_mul(length).expect("saved input width overflows")
}

fn main() {
	let root = std::env::var_os("RECIPE_TRIAL_DIRECTORY").map(PathBuf::from).unwrap_or_else(std::env::temp_dir);
	let bundle = root.join(format!("composition-candidate-{}.ogdl", std::process::id()));
	phase("setup");
	let data = recipe.data("data/numeric/sample_subfolders").target("target");
	let model = recipe.model()
		.layer(16)
		.moe(1, [
			layer(8),
			layer(24),
			layer(8),
			layer(24),
			layer(8),
			layer(24),
		])
		.layer(1)
		.loss(mse);
	phase("training");
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(17)
		.epochs(1)
		.log(all)
		.fp(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
	phase("inference");
	let input = vec![0.0; input_width(&bundle)];
	let output = recipe.infer(&bundle, &input);
	assert!(!output.is_empty());
	assert!(output.iter().all(|value| value.is_finite()));
	std::fs::remove_file(&bundle).unwrap_or_else(|error| panic!("cannot remove {}: {error}", bundle.display()));
}
```
backend=nv4
